ulog.cpp

```diff
@@ -0,0 +1,429 @@
+            return false;
+        }
+    } else {
+        const uint16_t diff = _expected_seq - seq;
+        if (diff < (UINT16_MAX / 2)) {
+            /* re-transmission */
+            return false;
+        }
+    }
+
+    *drop = true;
+    _expected_seq = seq + 1;
+    return true;
+}
+
+void ULog::_logging_data_process(mavlink_logging_data_t *msg)
+{
+    bool drops = false;
+
+    if (!_logging_seq(msg->sequence, &drops))
+        return;
+
+    /* Waiting for ULog header? */
+    if (_waiting_header) {
+        const uint8_t magic[] = ULOG_MAGIC;
+
+        if (msg->length < ULOG_HEADER_SIZE) {
+            /* This should never happen */
+            log_error("ULog header is not complete");
+            return;
+        }
+
+        if (memcmp(magic, msg->data, sizeof(magic))) {
+            log_error("Invalid ULog Magic number");
+            return;
+        }
+
+        _buffer_partial_len = ULOG_HEADER_SIZE;
+        memcpy(_buffer_partial, msg->data, ULOG_HEADER_SIZE);
+
+        memmove(msg->data, &msg->data[ULOG_HEADER_SIZE], msg->length);
+        msg->length -= ULOG_HEADER_SIZE;
+        _waiting_header = false;
+    }
+
+    if (drops) {
+        _logging_flush();
+
+        _buffer_len = 0;
+        _buffer_index = 0;
+        _waiting_first_msg_offset = true;
+    }
+
+    /*
+     * Do not cause a buffer overflow, it should only happens if a ULog message
+     * don't fit in _msg_buffer
+     */
+    if ((_buffer_len + msg->length) > BUFFER_LEN) {
+        log_warning("Buffer full, dropping everything on buffer");
+
+        _buffer_len = 0;
+        _waiting_first_msg_offset = true;
+    }
+
+    /*
+     * ULog message fits on _buffer but it need move all valid data to
+     * the being of buffer.
+     */
+    if ((_buffer_index + _buffer_len + msg->length) > BUFFER_LEN) {
+        memmove(_buffer, &_buffer[_buffer_index], _buffer_len);
+        _buffer_index = 0;
+    }
+
+    uint8_t begin = 0;
+
+    if (_waiting_first_msg_offset) {
+        if (msg->first_message_offset == NO_FIRST_MSG_OFFSET) {
+            /* no useful information in this message */
+            return;
+        }
+
+        _waiting_first_msg_offset = false;
+        begin = msg->first_message_offset;
+    }
+
+    if (!msg->length)
+        return;
+
+    msg->length = msg->length - begin;
+    memcpy(&_buffer[_buffer_index + _buffer_len], &msg->data[begin], msg->length);
+    _buffer_len += msg->length;
+    _logging_flush();
+}
+
+bool ULog::_logging_flush()
+{
+    while (_buffer_partial_len) {
+        const ssize_t r = write(_file, _buffer_partial, _buffer_partial_len);
+        if (r == 0 || (r == -1 && errno == EAGAIN))
+            return true;
+        if (r < 0) {
+            log_error_errno(errno, "Unable to write to ULog file: (%m)");
+            return false;
+        }
+
+        _buffer_partial_len -= r;
+        memmove(_buffer_partial, &_buffer_partial[r], _buffer_partial_len);
```

Comment:
Why not use this `r` as an offset to write?
Then you'd have something like:

```cpp
ssize_t r = 0;
while (_buffer_partial_len) {
    r = write(_file, _buffer_partial + r, _buffer_partial_len);
[...]
```

And you could get rid of this `memmove` altogether?